0011-rendezvous-hashing.md

```diff
@@ -0,0 +1,311 @@
+
+```
+message ShardingBlobAccessConfiguration {
+  message Shard {
+    // unchanged
+    BlobAccessConfiguration backend = 1;
+    // unchanged
+    uint32 weight = 2;
+  }
+  // NEW:
+  // Shards identified by a key within the context of this sharding
+  // configuration. The key is a freeform string which describes the identity
+  // of the shard in the context of the current sharding configuration.
+  // Shards are chosen via Rendezvous hashing based on the digest, weight and
+  // key of the configuration.
+  //
+  // When removing a shard from the map it is guaranteed that only blobs
+  // which resolved to the removed shard will get a different shard. When
+  // adding shards there is a weight/total_weight probability that any given
+  // blob will be resolved to the new shards.
+  map<string, Shard> shards = 2;
+  // NEW:
+  message Legacy {
+    // Order of the shards for the legacy schema. Each key here refers to
+    // a corresponding key in the 'shard_map' or null for drained backends.
+    repeated string shard_order = 1;
+    // Hash initialization seed used for legacy schema.
+    uint64 hash_initialization = 2;
+  }
+  // NEW:
+  // A temporary legacy mode which allows clients to use storage backends which
+  // are sharded with the old sharding topology implementation. Consumers are
+  // expected to migrate in a timely fashion and support for the legacy schema
+  // will be removed by 2025-12-31.
+  Legacy legacy = 3;
+}
+```
+
+## Migration Instructions
+
+If you are fine with dropping your data and repopulating your cache you can
+start using the new sharding algorithm right away. For clarity we have an
+example on how to perform a non-destructive migration.
+
+Given the following configuration showing an unmirrored sharding topology of two
+shards:
+
+```jsonnet
+{
+  ...
+  contentAddressableStorage: {
+    sharding: {
+      hashInitialization: 11946695773637837490,
+      shards: [
+        {
+          backend: {
+            grpc: {
+              address: 'storage-0:8980',
+            },
+          },
+          weight: 1,
+        },
+        {
+          backend: {
+            grpc: {
+              address: 'storage-1:8980',
+            },
+          },
+          weight: 1,
+        },
+      ],
+    },
+  },
+  ...
+}
+```
+
+We first want to non-destructively represent the exact same topology with the
+new configuration.
+
+```jsonnet
+{
+  ...
+  contentAddressableStorage: {
+    sharding: {
+      shards: {
+        "0": {
+          backend: { grpc: { address: 'storage-0:8980' } },
+          weight: 1,
+        },
+        "1": {
+          backend: { grpc: { address: 'storage-1:8980' } },
+          weight: 1,
+        },
+      },
+      legacy: {
+        shardOrder: ["0", "1"],
+        hashInitialization: 11946695773637837490,
+      },
+    },
+  },
+  ...
+}
+```
+
+The legacy parameter will make buildbarn internally use the old sharding
```

Comment:
Probably worth mentioning that this step is optional? As in, if you are willing to take downtime/lose cache contents, then it's all right to skip this.